### ISSUE TITLE
Extend Value type. 

### ISSUE THREAD

**Issue by tolysz** [bos/aeson#323]
Hi,

I know we had this discussion a couple of times already; but I really try to use `aeson`

    data Series = Empty
            | Value Encoding
            deriving (Typeable)

I see that `Empty` is only used for arrays; but I would like to have some kind of indication at the `Value` level that the result should be serialised as `Empty` in arrays and as omitted key in  `Object`s
(not just once `omitNothingFields` possibly different for each)

My proposal requires adding an Extra constructor to `Value` one which will represent removed/missing/one_to_be_removed key.

This extra constructor is needed if we want to be able to build more generic types.

say we have and `instance ToJSON a => Map Text a`
There is no way for `a` to tell `aeson` how it should serialise itself (i.e. `omitNothingFields` is not modifiable)  

My proposal is to add `Missing` constructor to `Value`  which will serialise same currently `Null` with `omitNothingFields=True` and the default for `Null` be `omitNothingFields=False`. 

This requires `.:??` extra operator which would return `Missing` (or `Omitted`) if there is no field and `Null` if there is `null`
and the `.:?` will behave as it currently does.

There are practical cases where it is useful; but you know that as you have `omitNothingFields`; but it not allow for the field to be `null` or `Omitted` depending on the actual value.

e.g.
   
     With a type `Maybe (Maybe a)` (Or any other type with the same universal property) I want to have:

    instance ToJSON a => Maybe (Maybe a)
    toJson Nothing = Omitted
      toJson (Just (Nothing)) = Null
      toJson (Just (Just a)) = toJson a

(I use it extensively https://github.com/tolysz/google-api ) I use patched `aeson-8.0.0.2` but I would like to port it upstream it compiles all libraries. 

Modification should be quite straightforward:
Add extra constructor; update encoders/decoders; update TH and Generics...

**Comment by phadej** [comment]
I don't think that adding a constructor to `Value` is good idea. There are various problems in your request, you try to solve with it:

### Distinguishing between missing and `Null` value:

with next release, there is `(.:!) :: FromJSON a => Object -> Text -> Parser (Maybe a)` which you can use to parse field into `Maybe (Maybe b)`. The `Nothing` will be missing value, `Just Nothing` would be `Null` and `Just (Just x)` present value.

### `instance ToJSON a => Map Text a`

Omitting some values from `Map` would break `toJSON . fromJSON` identity. I personally don't like that,

### Records

If there is the case when some fields should be omitted, but some serialised as `null`, then I'd rather implement own `Generics` or `TH` method

```hs
instance ToJSON MyRecord where
  toJSON = gToJSONWithMissing
```
where
`gToJSONWithMissing` would handle `Data.Possible` values, as vanilla generic encoder treats `Maybe`.

I don't see the need to patch `aeson` to provide this functionality.

**Comment by tolysz** [comment]
Adding extra `Value` is needed to be able to represent `JSON` as `Value`.
Currently there is no way, having only Value to know if we want to omit a given key in a map or display `null` there that information is not encoded in the type. You need to pass this value on the construction for a given type.

`toJSON . fromJSON` Can not get any more broken, as my approach is preserving more information.
More you can preserve the original encoded `text` i.e. keep `null` and not introduce `null` if you use the right type and want to have this functionality.

`(.:!)` looks promising for reading.

The only problem now is the serialisation, if I have a `Map Text Value` how do I know which keys to remove? Currently you need to specify that you remove all (`Null`) or none. And this shape is visible in the toJSON (via Value). So say I serialize `Possible a` to `Maybe Maybe a` there is no way to encode it as Value and have the information if we want to keep it or remove it.

The poof that it is needed is the `omitNothingFields` option; and I do not see how to have `omitMissingFields` and `keepNullas_null` without adding extra constructor.

**Comment by phadej** [comment]
I don't understand:

> Adding extra Value [constructor?] is needed to be able to represent JSON as `Value`.

AFAIK you can represent all valid JSON values in `Value`.

Totally different problem is how to represent different serialisation/deserialisation schemes, i.e. maps `a -> Value`. This library has made a choice to represent them directly. Thus missing some special cases, which in turn are handled by e.g. `omitNullaryFields`.

You can have `class ToJSONField a where toJSONField :: a -> Maybe Value`, if `omitNullaryFields` is to coarse, and use that in your serialisation implementation.

---

`fromJSON . toJSON`, how you approach is preserving more information? If some key is removed, where from `fromJSON` can know if it was ever present? Don't mix up serialisation of `Map` (arbitrary keys) and record types (known set of keys).

If you have `Map Text Value` it will be serialised so that all keys are in present resulting JSON value. Don't mix up serialisation of `Map` and arbitrary record types!

---

I repeat myself: you can role up your own serialisation mechanism, you don't need to alter provided one. Your need feels to be *very* specific. 

Using `generics-sop` it shouldn't be as difficult as using GHC generics. See https://github.com/well-typed/json-sop/blob/master/src/Generics/SOP/JSON.hs.

```hs
-- Not `ToJSONField`, not `ToJSON`.
myGtoJSON :: forall a. (Generic a, HasDatatypeInfo a, All2 ToJSONField (Code a))
           => JsonOptions -> a -> Value
```

Or you can use TH for your variant.

**Comment by tolysz** [comment]
> AFAIK you can represent all valid JSON values in Value.
The specs only say how it looks like when it is printed thus it is missing `epsilon` which has no representation.

Maybe I am missing something.
How can you represent serialisation of `ToJSON a => Map Text a` ? I can not, because `a` has to reduce to something of type `Value` and there is no way to know pepresentswhich keys to remove. 
Currently you can make your choice once -  `omitNullaryFields`, and either never or always have `null`.

Thus this extra value is used to mark cases where we have a `[("something1", Omitted), ("something2", Null)]` and want to get `{"something2":null}`and our container is generic (like Map), and we do not know it we should remove the `s'thng1` key until we have the value. 

What I am saying: I should be able to use `Maybe (Maybe a)` to read, but there is no generic way, to tell the container to remove or keep a given key from the sub-tree(when you compute to value on the second of the pair ).

The other practical side to this extra value, is that one could work with patches and say include only field which are different; But then it is crucial to be able to allow 'null' and omitingNullaryFields at the same time, as one could use the former to remove value and the later to leave it unchanged

**Comment by phadej** [comment]
@tolysz I already mentioned: don't mix up representation of **JSON value** (the `Value` type), and representation of **serialisation of Haskell values to `Value`** (i.e. `a -> Value` *or* some others). For `Map` case use e.g.

```hs
instance (JsonOmmittable a, MyToJSON a) => MyToJSON (Map Text a)
```

for your own `MyToJSON`.

**Comment by tolysz** [comment]
`omitNullaryFields` currently overloads `Null` value as 'null' and remove key from Object.
I need to be able to represent this decision on the `Value` level.
In the example:
> instance (JsonOmmittable a, MyToJSON a) => MyToJSON (Map Text a)
the `MyToJSON` can not be omitted. in some containing it Map without this extra constructor in `Value` which `MyToJSON` would return to signalise it want to emit empty representation.

The extra constructor is like '0' in mathematics it is not very useful in the actual representation; As there is no need show "0 rubber ducks" as physical objects but it does not mean it is not there.  

>instance ToJSON MyRecord where
>  toJSON = gToJSONWithMissing

will not work arbitrary deep without this extra constructor;
With it we might do not even need: `omitNullaryFields`, as it is a hack to store this extra constructor in the same place as `Null` for type `Maybe`. So effectively the additional information which should be part of Value is stored in a function in a way you can not inspect.

**Comment by phadej** [comment]
As a side note. Please, check the grammar and markdown formatting. I have troubles understanding you. 

Why creating and using
```haskell
class TolyczToJSON a where
  tolyczToJSON :: a -> Maybe Value

tolyczEncode :: TolyczToJSON a => a -> ByteString
```
is not an option for you?

**Comment by bergmark** [comment]
I've read this a couple of times but unfortunately I don't see what benefit modifying `Value` to make it contain more information than a json value can contain is a good idea, or what problem it solves compare d to other methods @phadej has suggested.

**Comment by bergmark** [comment]
I don't think this is the way to move forward, sorry.

**Comment by tolysz** [comment]
Well, the presented approach might be good, but I already know how to manually write instances for whatever types.
The problem is when we want to derive it. The Value (Top) would need to know somehow if the sub component wants to be omitted or not. currently we have to manually say it.
This approach (generic) forces either to omit all or none, otherwise one needs to manually write all instances.
The additional constructor(of Value) allow the sub component decide how it want to present itself.

Basically, what we have currently, is an the additioal proposed constructor switching placess with `Null ` constructor depending on the ommit...flag setting. 
Thus there will be no way of representing both at the same time in some generic way (universal property - as we pipe everything via Value proves that we have to have one more constructor)
